taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -468,43 +471,56 @@ func (self *Contract) Run(ctx vm.CallFrame, evm *vm.EVM) ([]byte, error) {
 			fmt.Println("Unable to parse getValidators input args: ", err)
 			return nil, err
 		}
-
-		res := self.getValidators(args)
-		return method.Outputs.Pack(res.Validators, res.End)
+		return method.Outputs.Pack(self.getValidators(args))
 
 	case "getDelegations":
 		var args sol.GetDelegationsArgs
 		if err = method.Inputs.Unpack(&args, input); err != nil {
 			fmt.Println("Unable to parse getDelegations input args: ", err)
 			return nil, err
 		}
-
-		res := self.getDelegations(args)
-		return method.Outputs.Pack(res.Delegations, res.End)
+		return method.Outputs.Pack(self.getDelegations(args))
 
 	case "getUndelegations":
 		var args sol.GetUndelegationsArgs
 		if err = method.Inputs.Unpack(&args, input); err != nil {
 			fmt.Println("Unable to parse getUndelegations input args: ", err)
 			return nil, err
 		}
-
-		res := self.getUndelegations(args)
-		return method.Outputs.Pack(res.Undelegations, res.End)
+		return method.Outputs.Pack(self.getUndelegations(args))
 	}
 
 	return nil, nil
 }
 
-func (self *Contract) DistributeRewards(rewardsStats *rewards_stats.RewardsStats, feesRewards *FeesRewards) {
+func (self *Contract) DistributeRewards(author *common.Address, rewardsStats *rewards_stats.RewardsStats, feesRewards *FeesRewards) {
 	// When calling DistributeRewards, internal structures must be always initialized
 	self.lazy_init()
 
 	// Calculates number of tokens to be generated as block reward
 	blockReward := bigutil.Mul(self.amount_delegated, self.yield_percentage)
 	blockReward = bigutil.Div(blockReward, bigutil.Mul(Big100, self.blocks_per_year))
 
+	voteReward := bigutil.Big0
+	authorReward := bigutil.Big0
+	trxReward := blockReward
+	// We need to handle case for block 0
+	if rewardsStats.TotalVotesCount > 0 {
+		// Calculate propotion between votes and transactions
+		voteReward = bigutil.Div(bigutil.Mul(blockReward, VotesToTrasnactionsRatio), Big100)
+		trxReward = bigutil.Sub(blockReward, voteReward)
+
+		// In case block author included more than minimum required number of votes
+		if rewardsStats.BonusVotesCount > 0 {
+			tmpVoteReward := bigutil.Div(voteReward, big.NewInt(int64(rewardsStats.TotalVotesCount)))
+			authorReward = bigutil.Mul(tmpVoteReward, big.NewInt(int64(rewardsStats.BonusVotesCount)))
```

Comment:
also here, you can mul first and then div

Reply:
I agree with the first part, but do not understand second part as `(total_count - bonus_count)` will get different output here

Reply:
Explanation here https://github.com/Taraxa-project/taraxa-evm/pull/87/files#r902343389

Example, lets say that huge number's last digit is 3, bonus_count 2, total_count 10
3 * 2 / 10 = 0
3 * 8 / 10 = 2

Reply:
Btw this means that block author might actually receive 33% from votes reward (which is 80% from total block reward), so in case block author put into his block all votes, he might actually receive approx. 26% from overall block reward, which is pretty big number. Originally we did not want reward pbft block author at all(as we agreed that creating dag blocks/voting is more important & requires more work) so this seems little bit too much

Reply:
So either we are ok with this, or we decide that the theoretical reward for including bonus votes is too big. In such we could alternatively use something like this:

- lets say we specify that max_block_author_reward might be max 5%(could be whatever we specify) of block reward for including bonus votes
- theoretical max number of votes is 100, block author included 80. To reach consensus from 100 votes, there must 67 valid votes so anything more than that is counted as bonus vote (mox bonus votes count is 33) - block author included 13 bonus votes.
- he will receive 13/33 * 100 [%] from max_block_author_reward

Reply:
he can not get more then 7% as votes are 20% of all rewards and he can only get 1/3 of that